bin/xbps-repodb/purge.c

```diff
@@ -0,0 +1,284 @@
+/*- * Copyright (c) 2020 Duncan Overbruck <mail@duncano.de>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <strings.h>
+#include <errno.h>
+#include <unistd.h>
+#include <limits.h>
+#include <dirent.h>
+
+#include <xbps.h>
+
+#include "uthash.h"
+#include "defs.h"
+
+struct arch {
+	char arch[64];
+	struct arch *next;
+	struct xbps_repo *repo, *stage;
+};
+
+struct repo {
+	char path[PATH_MAX];
+	struct repo *next;
+	struct dirent **namelist;
+	int nnames;
+	struct arch *archs;
+};
+
+static struct repo *repos;
+
+static void
+add_repo(struct xbps_handle *xhp, const char *path)
+{
+	struct repo *repo = calloc(1, sizeof (struct repo));
+	if (repo == NULL) {
+		perror("calloc");
+		exit(EXIT_FAILURE);
+	}
+	xbps_strlcpy(repo->path, path, sizeof repo->path);
+	repo->namelist = NULL;
+	repo->nnames = 0;
+	repo->archs = NULL;
+	repo->next = repos;
+	repos = repo;
+	xbps_dbg_printf(xhp, "Scanning repository: %s\n", path);
+
+	repo->nnames = scandir(path, &repo->namelist, NULL, NULL);
+	if (repo->nnames == -1) {
+		perror("scandir");
+		exit(EXIT_FAILURE);
+	}
+
+	for (int i = 0; i < repo->nnames; i++) {
+		const char *name = repo->namelist[i]->d_name, *d;
+		if (*name == '.')
+			continue;
+		if ((d = strrchr(name, '-')) == NULL)
+			continue;
+		if (strcmp(d+1, "repodata") == 0) {
+			struct arch *arch = calloc(1, sizeof (struct arch));
+			if (arch == NULL) {
+				perror("calloc");
+				exit(1);
+			}
+			if ((size_t)(d-name) >= sizeof arch->arch) {
+				xbps_error_printf("invalid repodata: %s\n", name);
+				exit(1);
+			}
+			strncpy(arch->arch, name, d-name);
```

Comment:
`strncpy(3)` should use the size of the destination for `n`, i.e. `strncpy(arch->arch, name, sizeof(name));`, not a size depending on the source. I see you made sure it fits into `arch->arch` before but gcc or clang will probably complain?

Reply:
Hm... no, they don't complain. Probably gcc is smart enough to *see* that this is ok.

Reply:
This is used to copy only the architecture part from from `XXX-repodata`, n is used here as how many bytes to copy and not as limit, for the limit use case I would have used `xbps_strlcpy`.